File: Mage/Map/Cache/CacheOverlayTableCell.m
//
//  CacheOverlayTableCell.m
//  MAGE
//
//  Created by Brian Osborn on 1/11/16.
//  Copyright © 2016 National Geospatial Intelligence Agency. All rights reserved.
//

#import "CacheOverlayTableCell.h"
#import "CacheOverlays.h"
#import "MageConstants.h"
#import "MAGE-Swift.h"

@interface CacheOverlayTableCell()<UITableViewDelegate, UITableViewDataSource>
@property (strong, nonatomic) id<MDCContainerScheming> scheme;
@end

@implementation CacheOverlayTableCell

- (id)initWithStyle:(UITableViewCellStyle)style reuseIdentifier:(NSString *)reuseIdentifier scheme: (id<MDCContainerScheming>) containerScheme {
    self = [super initWithStyle:style reuseIdentifier:reuseIdentifier];
    if (self) {
        self.tableView = [[UITableView alloc]initWithFrame:CGRectZero style:UITableViewStylePlain];
        [self.tableView setSeparatorStyle:UITableViewCellSeparatorStyleNone];
        self.tableView.tag = 100;
        self.tableView.delegate = self;
        self.tableView.dataSource = self;
        self.scheme = containerScheme;
        [self addSubview:self.tableView];
    }
    return self;
}

 -(void)layoutSubviews {
    [super layoutSubviews];
    UITableView *subMenuTableView = (UITableView *) [self viewWithTag:100];
    subMenuTableView.frame = CGRectMake(0.2, 0.3, self.bounds.size.width, self.bounds.size.height);
}

- (void) configure {
    [self.tableView reloadData];
}

- (CGFloat) tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return 58.0f;
}

- (CGFloat) tableView:(UITableView *)tableView heightForFooterInSection:(NSInteger)section {
    return 0.0f;
}

- (UIView *) tableView:(UITableView *)tableView viewForFooterInSection:(NSInteger)section {
    return [[UIView alloc] initWithFrame:CGRectZero];
}

-(NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

-(NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return [self.overlay getChildren].count == 1 ? 1 : [self.overlay getChildren].count + 1;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    if (indexPath.row == 0 && [self.overlay getChildren].count != 1) {
        self.overlay.expanded = !self.overlay.expanded;
        [self.tableView reloadData];
        [self.mainTable reloadData];
    }
    [tableView deselectRowAtIndexPath:[tableView indexPathForSelectedRow] animated:NO];
}

-(UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"cacheOverlayCell"];
    if(cell == nil) {
       cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleSubtitle reuseIdentifier:@"cacheOverlayCell"];
    }
    cell.textLabel.textColor = [self.scheme.colorScheme.onSurfaceColor colorWithAlphaComponent:0.87];
    cell.detailTextLabel.textColor = [self.scheme.colorScheme.onSurfaceColor colorWithAlphaComponent:0.6];
    cell.backgroundColor = self.scheme.colorScheme.surfaceColor;
    cell.imageView.tintColor = self.scheme.colorScheme.primaryColorVariant;
    
    if ([self.overlay getChildren].count != 1  && indexPath.row == 0) {
        cell.textLabel.textColor = [self.scheme.colorScheme.onSurfaceColor colorWithAlphaComponent:0.87];
        cell.detailTextLabel.textColor = [self.scheme.colorScheme.onSurfaceColor colorWithAlphaComponent:0.6];

        CacheActiveSwitch *cacheSwitch = [[CacheActiveSwitch alloc] initWithFrame:CGRectZero];
        cacheSwitch.on = self.overlay.enabled;
        cacheSwitch.overlay = self.overlay;
        cacheSwitch.onTintColor = self.scheme.colorScheme.primaryColorVariant;
        [cacheSwitch addTarget:self action:@selector(activeChanged:) forControlEvents:UIControlEventTouchUpInside];
        cell.accessoryView = cacheSwitch;
        cell.textLabel.text = self.mageLayer ? self.mageLayer.name : [self.overlay getName];
        cell.detailTextLabel.text = [NSString stringWithFormat: @"%lu layer%@", (unsigned long)[self.overlay getChildren].count, [self.overlay getChildren].count == 1 ? @"" : @"s"];
        [cell.imageView setImage:[UIImage imageNamed:@"folder"]];
        
    } else {
        CacheOverlay *cacheOverlay = [self.overlay.getChildren objectAtIndex:[self.overlay getChildren].count == 1 ? indexPath.row : indexPath.row - 1];
        UIImage * cellImage = nil;
        NSString * typeImage = [cacheOverlay getIconImageName];
        if(typeImage != nil){
            cellImage = [UIImage imageNamed:typeImage];
        }
        cell.textLabel.text = [self.overlay getChildren].count == 1 ? self.mageLayer ? self.mageLayer.name : [self.overlay getName] : [cacheOverlay getName];
        cell.detailTextLabel.text = [cacheOverlay getInfo];
        
        if (cellImage != nil) {
            [cell.imageView setImage:cellImage];
        }

        CacheActiveSwitch *cacheSwitch = [[CacheActiveSwitch alloc] initWithFrame:CGRectZero];
        cacheSwitch.on = cacheOverlay.enabled;
        cacheSwitch.overlay = cacheOverlay;
        cacheSwitch.onTintColor = self.scheme.colorScheme.primaryColorVariant;
        [cacheSwitch addTarget:self action:@selector(childActiveChanged:) forControlEvents:UIControlEventTouchUpInside];
        cell.accessoryView = cacheSwitch;
    }

    return cell;
}

- (IBAction)activeChanged:(CacheActiveSwitch *)sender {
    
    CacheOverlay * cacheOverlay = sender.overlay;
    
    [cacheOverlay setEnabled:sender.on];
    
    BOOL modified = false;
    for(CacheOverlay * childCache in [cacheOverlay getChildren]){
        if(childCache.enabled != cacheOverlay.enabled){
            [childCache setEnabled:cacheOverlay.enabled];
            modified = true;
        }
    }
    
    if(modified){
        [self.tableView reloadData];
    }
    
    [self updateSelectedAndNotify];
}

- (IBAction)childActiveChanged:(CacheActiveSwitch *)sender {
    
    CacheOverlay * cacheOverlay = sender.overlay;
    CacheOverlay * parentOverlay = [cacheOverlay getParent];
    
    [cacheOverlay setEnabled:sender.on];
    
    BOOL parentEnabled = true;
    if(!cacheOverlay.enabled){
        parentEnabled = false;
        for(CacheOverlay * childOverlay in [parentOverlay getChildren]){
            if(childOverlay.enabled){
                parentEnabled = true;
                break;
            }
        }
    }
    if(parentEnabled != parentOverlay.enabled){
        [parentOverlay setEnabled:parentEnabled];
        [self.tableView reloadData];
    }
    
    [self updateSelectedAndNotify];
}

-(void) updateSelectedAndNotify{
    NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
    NSMutableArray * overlays = [[NSMutableArray alloc] init];
    CacheOverlays *cacheOverlays = [CacheOverlays getInstance];
    for(CacheOverlay * cacheOverlay in [cacheOverlays getOverlays]){

        BOOL childAdded = false;
        for(CacheOverlay * childCache in [cacheOverlay getChildren]){
            if(childCache.enabled){
                [overlays addObject:[childCache getCacheName]];
                childAdded = true;
            }
        }

        if(!childAdded && cacheOverlay.enabled){
            [overlays addObject:[cacheOverlay getCacheName]];
        }
    }
    [defaults setObject:overlays forKey:MAGE_SELECTED_CACHES];
    [defaults synchronize];
    dispatch_async(dispatch_get_main_queue(), ^{
        [cacheOverlays notifyListeners];
    });
}

@end

File: Mage/Map/Cache/CacheOverlayTableCell.swift
//
//  CacheOverlayTableCell.m
//  MAGE
//
//  Created by Brian Osborn on 1/11/16.
//  Copyright © 2016 National Geospatial Intelligence Agency. All rights reserved.
//
import Foundation
import UIKit

class CacheOverlayTableCell: UITableViewCell {
    var active: CacheActiveSwitch?
    var tableType: UIImageView?
    var name: UILabel?
    var overlay: CacheOverlay?
    var mageLayer: Layer?
    var mainTable: UITableView?
    weak var excludedListener: (any CacheOverlayListener)? // Used to prevent duplicate reloads when tapping toggles
    var tableView: UITableView
    
    var scheme: MDCContainerScheming?
    
    init(style: UITableViewCell.CellStyle, reuseIdentifier: String?, scheme: MDCContainerScheming) {
        self.tableView = UITableView(frame: .zero, style: .plain)
        self.tableView.tag = 100
        self.scheme = scheme
        
        super.init(style: style, reuseIdentifier: reuseIdentifier)

        self.tableView.dataSource = self
        self.tableView.delegate = self
        self.addSubview(self.tableView)
    }
    
    override func layoutSubviews() {
        super.layoutSubviews()
        let subMenuTableView = self.viewWithTag(100)
        subMenuTableView?.frame = CGRect(x: 0.2, y: 0.3, width: self.bounds.size.width, height: self.bounds.size.height)
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    func configure() {
        tableView.reloadData()
    }
    
    @objc func activeChanged(sender: CacheActiveSwitch) {
        guard let cacheOverlay = sender.overlay else { return }
        cacheOverlay.enabled = sender.isOn
        
        var modified = false
        for childCache in cacheOverlay.getChildren() {
            if childCache.enabled != cacheOverlay.enabled {
                childCache.enabled = cacheOverlay.enabled
                modified = true
            }
        }
        
        if modified {
            // Refresh the parent row for expanded layers (GeoPackage)
            tableView.reloadRows(at: rowIndexPathsForLayers(overlay: cacheOverlay), with: .none)
        }
        Task {
            await updateSelectedAndNotify()
        }
    }

    @objc func childActiveChanged(sender: CacheActiveSwitch) {
        guard let cacheOverlay = sender.overlay else { return }
        guard let parentOverlay = cacheOverlay.getParent() else { return }
        
        cacheOverlay.enabled = sender.isOn
        
        var parentEnabled = true
        if !cacheOverlay.enabled {
            parentEnabled = false
            for childCache in parentOverlay.getChildren() {
                if childCache.enabled {
                    parentEnabled = true
                    break
                }
            }
        }
        
        if parentEnabled != parentOverlay.enabled {
            parentOverlay.enabled = parentEnabled
            // Refresh the parent row for expanded layers (GeoPackage)
            if let parentRow = rowIndexPathForExpandableGroupIfPresent(overlay: parentOverlay) {
                tableView.reloadRows(at: [parentRow], with: .none)
            }
        }
        Task {
            await updateSelectedAndNotify()
        }
    }
    
    func updateSelectedAndNotify() async {
        var overlays: [String] = []
        var cacheOverlaysOverlays: [CacheOverlay] = []
        cacheOverlaysOverlays.append(contentsOf: await CacheOverlays.shared.getOverlays())
        
        for cacheOverlay in cacheOverlaysOverlays {
            var childAdded = false
            for childCache in cacheOverlay.getChildren() {
                if childCache.enabled {
                    overlays.append(childCache.cacheName)
                    childAdded = true
                }
            }
            
            if !childAdded && cacheOverlay.enabled {
                overlays.append(cacheOverlay.cacheName)
            }
        }
        
        UserDefaults.standard.selectedCaches = overlays
        await CacheOverlays.shared.notifyListenersExceptCaller(caller: excludedListener)
    }

    private func isExpandableGroup(_ overlay: CacheOverlay) -> Bool {
        overlay.getChildren().count > 1
    }

    private func rowIndexPathsForLayers(overlay: CacheOverlay) -> [IndexPath] {
        // Single-child overlays do not have an expand/collapse parent row; child lives at row 0.
        if !isExpandableGroup(overlay) {
            return [IndexPath(row: 0, section: 0)]
        }

        // Multi-child overlays use row 0 as the expandable parent and rows 1...n as children.
        return overlay.getChildren().indices.map { IndexPath(row: $0 + 1, section: 0) }
    }

    private func rowIndexPathForExpandableGroupIfPresent(overlay: CacheOverlay) -> IndexPath? {
        // Multi-child overlays have a parent row at index 0; single-child overlays do not.
        isExpandableGroup(overlay) ? IndexPath(row: 0, section: 0) : nil
    }

    private var layerToggleOnColor: UIColor {
        UIColor(named: "layerToggleOn") ?? .systemBlue
    }
}

extension CacheOverlayTableCell: UITableViewDataSource, UITableViewDelegate {
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        guard let overlay = overlay else { return 0 }
        return overlay.getChildren().count == 1 ? 1 : overlay.getChildren().count + 1
    }
    
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        58.0
    }
    
    func tableView(_ tableView: UITableView, heightForFooterInSection section: Int) -> CGFloat {
        0.0
    }
    
    func tableView(_ tableView: UITableView, viewForFooterInSection section: Int) -> UIView? {
        UIView(frame: .zero)
    }
    
    func numberOfSections(in tableView: UITableView) -> Int {
        1
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        guard let overlay = overlay else { return }
        if indexPath.row == 0 && isExpandableGroup(overlay) {
            overlay.expanded.toggle()
            tableView.reloadData()
            if let mainTable = mainTable {
                mainTable.reloadData()
            }
            tableView.deselectRow(at: indexPath, animated: false)
        }
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let identifier = "cacheOverlayCell"
        let cell = tableView.dequeueReusableCell(withIdentifier: identifier)
        ?? UITableViewCell(style: .subtitle, reuseIdentifier: identifier)
        
        cell.textLabel?.textColor = scheme?.colorScheme.onSurfaceColor.withAlphaComponent(0.87)
        cell.detailTextLabel?.textColor = scheme?.colorScheme.onSurfaceColor.withAlphaComponent(0.6)
        cell.backgroundColor = scheme?.colorScheme.surfaceColor
        cell.imageView?.tintColor = scheme?.colorScheme.primaryColorVariant
        
        if let overlay = overlay {
            let children = overlay.getChildren()
            let count = children.count
            
            if count != 1 && indexPath.row == 0 {
                // Parent row
                let cacheSwitch = CacheActiveSwitch(frame: .zero)
                cacheSwitch.isOn = overlay.enabled
                cacheSwitch.overlay = overlay
                cacheSwitch.onTintColor = layerToggleOnColor
                cacheSwitch.addTarget(self, action: #selector(activeChanged), for: .valueChanged)
                cell.accessoryView = cacheSwitch
                
                cell.textLabel?.text = (self.mageLayer != nil) ? self.mageLayer!.name : overlay.name
                cell.detailTextLabel?.text = "\(count) layer\(count == 1 ? "" : "s")"
                cell.imageView?.image = UIImage(systemName: "folder")
                
            } else {
                // Child row
                let childIndex = (count == 1) ? indexPath.row : indexPath.row - 1
                let cacheOverlay = children[childIndex]
                
                let imageName = cacheOverlay.iconImageName
                if let imageName = imageName {
                    cell.imageView?.image = UIImage(named: imageName)
                }
                
                cell.textLabel?.text = overlay.getChildren().count == 1 ? (self.mageLayer != nil ? self.mageLayer!.name : overlay.name) : cacheOverlay.name
                cell.detailTextLabel?.text = cacheOverlay.getInfo()
                
                let cacheSwitch = CacheActiveSwitch(frame: .zero)
                cacheSwitch.isOn = cacheOverlay.enabled
                cacheSwitch.overlay = cacheOverlay
                cacheSwitch.onTintColor = layerToggleOnColor
                cacheSwitch.addTarget(self, action: #selector(childActiveChanged), for: .valueChanged)
                cell.accessoryView = cacheSwitch
            }
        }
        return cell
    }
}
